File: 36_valid_sudoku/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        bool isValidSudoku(vector<vector<char>>& board) {
            // Columns
            for (int col = 0; col < 9; col++) {
                int appearances[9] = {0,0,0,0,0,0,0,0,0};

                for (int row = 0; row < 9; row++) {
                    if (board[row][col] == '.')
                        continue;
                    int number = (board[row][col] - '0') - 1;
                    appearances[number]++;
                    cout << (board[row][col] - '0') << endl;
                    if (appearances[number] > 1) {
                        return false;
                    }
                }
            }

            //Rows
            for (int row = 0; row < 9; row++) {
                int appearances[9] = {0,0,0,0,0,0,0,0,0};

                for (int col = 0; col < 9; col++) {
                    if (board[row][col] == '.')
                        continue;
                    int number = (board[row][col] - '0') - 1;
                    appearances[number]++;
                    cout << (board[row][col] - '0') << endl;
                    if (appearances[number] > 1) {
                        return false;
                    }
                }
            }

            return true;
        }
    };

File: 36_valid_sudoku/python/solution.py
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # empty squares are denoted by periods

        squares = [[[], [], []], [[], [], []], [[], [], []]]
        rows = [[] for num in range(0,9)]
        cols = [[] for num in range(0,9)]

        # check rows
        for row in range(0, 9):
            for col in range(0, 9):
                num = board[row][col]

                if rows[row].count(num) != 0:
                    return False
                elif cols[col].count(num) != 0:
                    return False
                elif squares[row // 3][col // 3].count(num) != 0:
                    return False
                
                rows[row].append(num)
                cols[col].append(num)
                squares[row // 3][col // 3].append(num)

        return True

